### PR TITLE
Enable tqdm toggling

### LINKED ISSUE #14889: [logging] unable to turn off tqdm logging
I'm writing a program in a notebook where I'm printing a results table for multiple models and I can't figure out how to turn off tqdm here as it downloads new models since its output breaks the table and adds a ton of unnecessary noise to the notebook's outputs.

So here is where the control is:

https://github.com/huggingface/transformers/blob/207594be81b8e5a8589c8b11c3b236924555d806/src/transformers/file_utils.py#L1890-L1898

Why does it turn off only if logging is `logging.NOTSET`?  `logging.NOTSET` is a special level. If it's `NOTSET` then the next handler's log level is used, so we end up falling back to the root's default log level which is WARNING. (detailed info: https://docs.python.org/3/library/logging.html#logging.Logger.setLevel)

In other words:

https://github.com/huggingface/transformers/blob/207594be81b8e5a8589c8b11c3b236924555d806/src/transformers/file_utils.py#L1897

will never be true.

So we need to choose a true level at which tqdm is enabled. If it's an info log level, then most likely something like:

`disable=bool(logging.get_verbosity() >= logging.WARNING)`?

Thank you

@LysandreJik, @sgugger 

-----------

a quick test for myself to check that tqdm can be turned off once we sort it out (assuming this model is not cached):

```
python -c "import transformers, logging; transformers.logging.set_verbosity_error(); \
transformers.AutoModel.from_pretrained('google/pegasus-pubmed')"
```
shouldn't log anything, but try `set_verbosity_info` first to ensure that the model isn't cached yet and kill it before it finishes.

### PR DESCRIPTION
**Pull request by jaketae**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/14889.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@stas00, @LysandreJik

### CODE REVIEW

**Comment by jaketae** [comment]
1. In `Datasets`, `tqdm_utils.py` functions are exposed at the upper level via imports in `__init__.py` (e.g. `datasets.utils.set_progress_bar_enabled` as opposed to ``datasets.utils.tqdm_utils.set_progress_bar_enabled`). It seems that this is not the case with utility functions in `transformers`. Do we want to keep functions like `datasets.utils.set_progress_bar_enabled` hidden at its current level?
2. Should toggling `set_progress_bar_enabled` affect the entire API, for instance the behavior of the `Trainer` class? This would mean replacing all occurrences of `from tqdm.auto import tqdm` with `from .utils.tqdm_utils import tqdm`.

**Review by stas00 — CHANGES_REQUESTED** [review]
Fantastic work, @jaketae!

- Let's also document this new feature in `docs/source/main_classes/logging.mdx` as this is a logging feature and I'd look for this first.

- perhaps also add a note in `def http_get` to reference  `transformers.utils.tqdm_utils.set_progress_bar_enabled` if someone is seeking out the turn off tqdm at the source - since we removed that code that wasn't doing anything.

**Comment by stas00** [comment]
> 1. In `datasets`, `tqdm_utils.py` functions are exposed at the upper level via imports in `__init__.py` ([link](https://github.com/huggingface/datasets/blob/ccc28d4141006924af85c2d940d82d71da4c6ec0/src/datasets/utils/__init__.py#L41)). It seems that this is not the case with utility functions in `transformers`. Do we want to keep functions like `transformers.utils.set_progress_bar_enabled` hidden at its current level?

I don't think there is a need for that. This will be very rarely used.

> 2. Should toggling `set_progress_bar_enabled` affect the entire API, for instance the behavior of the `Trainer` class? This would mean replacing all occurrences of `from tqdm.auto import tqdm` with `from .utils.tqdm_utils import tqdm`.

No. the turning off is advanced use, so we shouldn't turn tqdm anywhere since it has a very specific goal.

And those who don't want it can now turn it off at will.

File: tests/test_tqdm_utils.py
Comment (stas00) [inline tests/test_tqdm_utils.py:6]:
```suggestion
from transformers import AutoModel
from transformers.utils.tqdm_utils import set_progress_bar_enabled
from transformers.testing_utils import  require_torch

@require_torch
```

pytorch tests need to be designated as such, since otherwise tf/flax CI will fail as it doesn't have pytorch.

you will probably will need to run `make fixup` and push to keep the autoformatter happy.

Comment (stas00) [reply tests/test_tqdm_utils.py:6]: I think my 2 suggestions overlap, I hope it's easy to sort out.

File: tests/test_tqdm_utils.py
Comment (jaketae) [reply tests/test_tqdm_utils.py:16]: Ah, I was actually wondering if there is a smaller model to minimize CI overhead! Thanks for this suggestion.

**Comment by jaketae** [comment]
@stas00 Thanks for the review. I've updated the test and the docs/docstrings. Let me know if there is anything else that needs to be done!

File: docs/source/main_classes/logging.mdx
Comment (stas00) [inline docs/source/main_classes/logging.mdx:59]:
```suggestion
By default, `tqdm` progress bars will be displayed to show progress during model download, training in `Trainer` and various other places. [`transformers.utils.tqdm_utils.set_progress_bar_enabled`] can be used to suppress or unsuppress this behavior. 
```

it's not just download progress...

and added auto-link to the doc

**Review by stas00 — APPROVED** [review]
Looks great, @jaketae!

Just one last nit

File: docs/source/main_classes/logging.mdx
Comment (jaketae) [reply docs/source/main_classes/logging.mdx]: As it is now, toggling `set_progress_bar_enabled` won't change the behavior of the progress bar in the `Trainer` class. `Trainer` still uses `tqdm.auto.tqdm`.

https://github.com/huggingface/transformers/blob/226e74b610bd8d30c2faba627f245d693ced7b55/src/transformers/trainer.py#L33

For `set_progress_bar_enabled` also take affect in `Trainer`, the above import line should be replaced with `from .utils.tqdm_utils import tqdm`. This is in line with my previous inquiry:

> Should toggling `set_progress_bar_enabled` affect the entire API, for instance the behavior of the Trainer class? This would mean replacing all occurrences of from `tqdm.auto import tqdm` with `from .utils.tqdm_utils import tqdm`.

To my ears, the edit misleadingly suggests otherwise. What do you think? (I should have asked before committing...)

**Comment by stas00** [comment]
@jaketae, you left a comment and then deleted it - I still got it in the email notification.

I think what you were saying is totally correct, so perhaps add your notes to the logging doc. That's that model download is already controlled by `set_progress_bar_enabled`, and anywhere else where `tqdm.auto.tqdm` is used to make it controllable in the same way the user has to swap it for `from transformers.utils.tqdm_utils import tqdm`

File: docs/source/main_classes/logging.mdx
Comment (stas00) [reply docs/source/main_classes/logging.mdx]: I think what you are saying is totally correct, so perhaps add your notes to the logging doc. That's that model download is already controlled by set_progress_bar_enabled, and anywhere else where tqdm.auto.tqdm is used to make it controllable in the same way the user has to swap it for from transformers.utils.tqdm_utils import tqdm

File: tests/test_tqdm_utils.py
Comment (LysandreJik) [inline tests/test_tqdm_utils.py:22]: Could this download a configuration file instead, to be framework agnostic?

And this test could go into `tests/test_logging.py` so as to not add a new file.

**Review by LysandreJik — APPROVED** [review]
This looks good to me, thank you for your contribution @jaketae! 

Also cc @lhoestq as inspiraiton was taken from datasets.

Please wait for Sylvain's review before merging.

File: docs/source/main_classes/logging.mdx
Comment (sgugger) [inline docs/source/main_classes/logging.mdx:57]: `set_progress_bar_enabled` does not mean anything to me. I understand it's been taken from datasets so we may want to keep it for consistency between the libraries but can we add `disable_progress_bars()` and `enable_progress_bars()` as convenience methods with better names?

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

File: docs/source/main_classes/logging.mdx
Comment (jaketae) [reply docs/source/main_classes/logging.mdx:57]: I agree that the naming might not be as intuitive. Perhaps worthy of note, however, is that `disable_progress_bar()` is a [deprecated pattern](https://github.com/huggingface/datasets/blob/ccc28d4141006924af85c2d940d82d71da4c6ec0/src/datasets/utils/tqdm_utils.py#L83-L95) in `datasets`. 

```
@deprecated("Use set_progress_bar_enabled(False) instead.")
def disable_progress_bar():
    """Disable tqdm progress bar.

    .. deprecated:: 1.12.2
        Use set_progress_bar_enabled(False) instead.
```

Part of me is concerned that this might bring further confusion, but if you think this is a non-issue, I'll go ahead and add the two functions!

Comment (sgugger) [reply docs/source/main_classes/logging.mdx:57]: @lhoestq Why is this pattern deprecated?

Comment (lhoestq) [reply docs/source/main_classes/logging.mdx:57]: It comes from this PR: https://github.com/huggingface/datasets/pull/3031

Maybe let's discuss what should be the best API for `transformers` and `datasets` for consistency ? also cc @mariosasko

Comment (sgugger) [reply docs/source/main_classes/logging.mdx:57]: I personally would prefer having the two methods mentioned above.

File: src/transformers/file_utils.py
Comment (stas00) [reply src/transformers/file_utils.py:1908]: I suggested that Jake adds it here, because if you look at the original it had a way (albeit broken) of changing the behavior. 

So we aren't changing anything, just replacing the old broken behavior with a new working one and documenting that new way.

File: tests/test_tqdm_utils.py
Comment (stas00) [reply tests/test_tqdm_utils.py:22]: and then you can probably remove `require_torch`, once switched to `AutoConfig.from_pretrained(...)`

File: src/transformers/file_utils.py
Comment (sgugger) [reply src/transformers/file_utils.py:1908]: The documentation is only in the end function, so if you want to document it so a user can find it, it would need to go in every function that ends up calling this one. I'd keep things simple and just remove this.

Comment (stas00) [reply src/transformers/file_utils.py:1908]: Here is how I think about it: `http_get` is the function that invokes tqdm, and not its callers. Therefore a user tracing the tqdm call will find that `http_get` calls `tqdm`, and read the doc that explains how to control this behavior.

If this note is removed now the user will have no idea how to control this behavior, unless they think to go to the logging doc and read it there, which they may or may not think of it, since tqdm doesn't use logging.

Comment (sgugger) [reply src/transformers/file_utils.py:1908]: What you're describing would be better achieved with a comment in the code, not the dosctring.

Comment (stas00) [reply src/transformers/file_utils.py:1908]: That works. Let's do that.

File: docs/source/main_classes/logging.mdx
Comment (mariosasko) [reply docs/source/main_classes/logging.mdx:57]: As explained in the linked PR, my idea was to align the naming with `set_caching_enabled/is_caching_enabled` in `datasets`.

@lhoestq   
I guess we can add `enable_progress_bars` and remove the deprecation notice from `disable_progress_bar`, but then we should also do the same for the caching mechanism? WDYT? Also, I'm not a big fan of having two ways to do the same thing (`set_progress_bar_enabled(True)` and `enable_progress_bars()`).

Comment (stas00) [reply docs/source/main_classes/logging.mdx:57]: Perhaps `tqdm_progress_bars_toggle(True/False)` 

I added `tqdm` as since if it's always tqdm that is being used then it's really tqdm that we fiddle with.

but `progress_bars_toggle(True/False)`  is fine too.

Comment (sgugger) [reply docs/source/main_classes/logging.mdx:57]: I'm fine with one method if it's better named, for instance the last suggestion of @stas00 woks for me too. It's just that`set_progress_bar_enabled` is very hard to parse IMO.

Comment (stas00) [reply docs/source/main_classes/logging.mdx:57]: thinking more, linguistically "toggle" means to switch to the opposite state, so the name I proposed is a not the best.

Comment (stas00) [reply docs/source/main_classes/logging.mdx:57]: Perhaps having these 2 is the most intuitive? `tqdm_progress_bars_enable()` and `tqdm_progress_bars_disable()`

Comment (lhoestq) [reply docs/source/main_classes/logging.mdx:57]: I'm fine if we go back to using the `disable_progress_bar` in `datasets`.

Note that is doesn't have an `s` in the end though, contrary to your first suggestion @sgugger - so I would prefer if the one in `transformers` doesn't have an `s` either.

Then in this case @mariosasko we should have the `enable_progress_bar` function as well, and probably align the corresponding methods for caching indeed

Comment (sgugger) [reply docs/source/main_classes/logging.mdx:57]: Works for me!

**Comment by jaketae** [comment]
Thanks everyone for the comments! I've distilled the discussion into a to-do list, which I'll update as I go.

- [ ] Use `AutoConfig` to make the test framework-agnostic
- [ ] Move the test into `test_logging.py` instead of creating a separate new file
- [ ] Create `enable_progress_bar` and `disable_progress_bar`
- [ ] Create issue on `datasets` side to unify API/de-deprecate `disable_progress_bar`
- [ ] Move doctoring to comment

File: tests/test_tqdm_utils.py
Comment (jaketae) [reply tests/test_tqdm_utils.py]: To keep in line with the structure of the test directory (i.e. having `tqdm` tests in `test_logging,py`), does it also make sense to move `tqdm`-related code to `utils/logging.py` instead of keeping them in `utils/tqdm_utils.py`?

Comment (stas00) [reply tests/test_tqdm_utils.py]: I think it's a great idea and consistent with logging tests and docs.

But perhaps secure another agreement from @sgugger or @LysandreJik before you act on it.

Comment (jaketae) [reply tests/test_tqdm_utils.py]: I made this a tentative item on the to-do, but I'll wait for more feedback!

Comment (LysandreJik) [reply tests/test_tqdm_utils.py]: Yes, this sounds good to me! Having everything in `utils/logging.py` sounds great.

**Review by sgugger — APPROVED** [review]
Thanks a lot for iterating on this with us!

**Pull request merged**
